pkg/controller/core/resourceflavor_controller.go

```diff
@@ -18,36 +18,72 @@ package core
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/go-logr/logr"
 	"k8s.io/klog/v2"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 
 	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
 	"sigs.k8s.io/kueue/pkg/cache"
 	"sigs.k8s.io/kueue/pkg/queue"
 )
 
+const ResourceFlavorFinalizerName = "kueue.io/resourceflavor-protection"
+
 // ResourceFlavorReconciler reconciles a ResourceFlavor object
 type ResourceFlavorReconciler struct {
 	log      logr.Logger
 	qManager *queue.Manager
 	cache    *cache.Cache
+	client   client.Client
 }
 
-func NewResourceFlavorReconciler(qMgr *queue.Manager, cache *cache.Cache) *ResourceFlavorReconciler {
+func NewResourceFlavorReconciler(client client.Client, qMgr *queue.Manager, cache *cache.Cache) *ResourceFlavorReconciler {
 	return &ResourceFlavorReconciler{
 		log:      ctrl.Log.WithName("resourceflavor-reconciler"),
 		cache:    cache,
+		client:   client,
 		qManager: qMgr,
 	}
 }
 
-//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch;update;delete
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors/finalizers,verbs=update
 
 func (r *ResourceFlavorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	// Nothing to do here.
+	var flavor kueue.ResourceFlavor
```

Comment:
Reconcile is currently only triggered by events in the ResourceFlavor itself. However, we need to react to updates to ClusterQueues.

I would use this high level approach:

1. Add the finalizer in a mutation webhook for creation.
2. Only remove the finalizer when there is DeletionTimestamp and nothing is using the flavor.

Reply:
> Reconcile is currently only triggered by events in the ResourceFlavor itself. However, we need to react to updates to ClusterQueues.
> 
Can you provide some cases when resourceFlavor should react to clusterQueue. IMO, clusterQueue should react to resourceFlavor instead.

> I would use this high level approach:
> 
> 1. Add the finalizer in a mutation webhook for creation.

Sorry, I can't tell the difference between 1) add finalizer in a webhook or 2) in the create predicate. Reduce the complexity in `Reconcile`? What's more, in the future, Is it possible to disable webhook for tests etc.? 
> 2. Only remove the finalizer when there is DeletionTimestamp and nothing is using the flavor.

As the design now.

Reply:
what do you mean by create predicate?

Adding the finalizer in Reconcile means that there is a time when the ResourceFlavor doesn't have the finalizer, which could lead to problems. Doing it in the webhook guarantees that it always has the finalizer.

> Can you provide some cases when resourceFlavor should react to clusterQueue

Let's say a user tries to delete a ResourceFlavor that is currently in use. This adds a deletionTimestamp. Reconcile runs and doesn't remove the finalizer.

Next, the user deletes all the ClusterQueues that use the ResourceFlavor. What causes Reconcile for the ResourceFlavor to run again?

Reply:
> what do you mean by create predicate?
> 

I mean the `func (r *ResourceFlavorReconciler) Create(e event.CreateEvent) bool` below
> Adding the finalizer in Reconcile means that there is a time when the ResourceFlavor doesn't have the finalizer, which could lead to problems. Doing it in the webhook guarantees that it always has the finalizer.
> 

SG, but we don't have defaulting webhook for ResourceFlavor now or let's add this later?

> > Can you provide some cases when resourceFlavor should react to clusterQueue
> 
> Let's say a user tries to delete a ResourceFlavor that is currently in use. This adds a deletionTimestamp. Reconcile runs and doesn't remove the finalizer.
> 
> Next, the user deletes all the ClusterQueues that use the ResourceFlavor. What causes Reconcile for the ResourceFlavor to run again?

When we found ResourceFlavor is still in use, we'll return error and requeue infinitely, but the problem here is the retry duration may grow quite long. FYI ClusterQueue may also watching ResourceFlavor events. E.g. ClusterQueue is inactive now, when the required ResourceFlavor created, ClusterQueue Controller will be triggered to update the clusterQueue's status to active.

Reply:
> we'll return error and requeue infinitely

That sounds wasteful, specially since the implementation of `FlavorInUse` iterates through the entire cache. Btw, you should be using a lock, which will make things worse.

> FYI ClusterQueue may also watching ResourceFlavor events. 

That's fine.

Reply:
I wouldn't leave it for a follow up. Having finalizers introduces risk. We need to make sure we are handling the removal correctly.